drivers: display: features for ILI9341 driver

NickIOT72 opened the pull request:
**Overview**

This PR adds some features on `zephyr/driver/display` files related with `ili9xxx` header

**Features Implemented**

- modification on `display_ili9xxx.c` and `display_ili9xxx.h` files

- add condition to select BGR/RGB color order depending on pixel format

- change CMD_SET for ILI9341


**Files Modified/Added**

`zephyr/drivers/display/display_ili9xxx.c` - changes made on MADCTL register

`zephyr/drivers/display/display_ili9xxx.h` - Comment modified

`zephyr/dts/bindings/display/ilitek,ili9xxx-common.yaml` - Added BGR565 value on enum

`zephyr/include/zephyr/dt-bindings/display/ili9xxx.h` - added `ILI9XXX_PIXEL_FORMAT_BGR565` on pixel format


Fixes #99134

NickIOT72 committed:
drivers: display: issue with configuration on ILI9341

ILI9341 is not deploying correctly the display sample.
The screen looks mirrored vertically and the color
doesn't match with the sequence expected. To fix it,
change the Memory Access Control configuration data.

Signed-off-by: Nicolas Moreno <niko722795@gmail.com>

JarmouniA commented on dts/bindings/display/ilitek,ili9xxx-common.yaml:
```diff
@@ -20,6 +20,7 @@ properties:
     enum:
       - 0 # RGB565
       - 1 # RGB888
+      - 2 # BGR565
```
This property should actually be inhereted from `lcd-controller.yaml`

JarmouniA replied:
Remove the property and include `lcd-controller.yaml`.

danieldegrasse commented on drivers/display/display_ili9xxx.h:
```diff
@@ -59,7 +59,7 @@
 enum madctl_cmd_set {
 	CMD_SET_1,	/* Default for most of ILI9xxx display controllers */
-	CMD_SET_2,	/* Used by ILI9342c */
+	CMD_SET_2,	/* Used by ILI9342c & ILI9341 */
```
Have you verified that display orientation works with this change? Assuming that is why it was made

JarmouniA replied:
Maybe we should add the delay between the squares in the sample, seems useful.

NickIOT72 committed:
drivers: display: features for ILI9341 driver

Replaced display-controller.yaml with lcd-controller.yaml
Deleted pixel format property, replacing it with the
property on lcd-controller.yaml. Replace ILI9XXX RGB macro
with PANEL RGB macro. Also, added condition to verify pixel
format is RGB565, BRG565 or RGB888, otherwise it will show error
Replaced <zephyr/dt-bindings/display/ili9xxx.h> with
<zephyr/dt-bindings/display/panel.h> and
ILI9XXX_PIXEL with PANEL_PIXEL in some Devicetrees
and files that contained both elements. Fixed some
script sintax. Deleted drivers/display/display_ili9xxx.c

Signed-off-by: Nicolas Moreno <niko722795@gmail.com>

NickIOT72 commented on drivers/display/display_ili9xxx.c:
```diff
@@ -324,7 +324,8 @@ static int ili9xxx_set_orientation(const struct device *dev,
 
 	int r;
-	uint8_t tx_data = ILI9XXX_MADCTL_BGR;
+	uint8_t tx_data = config->pixel_format == PIXEL_FORMAT_BGR_565
```
Done

JarmouniA commented on drivers/display/display_ili9xxx.c:
```diff
@@ -394,6 +395,8 @@ static int ili9xxx_configure(const struct device *dev)
 	/* pixel format */
 	if (config->pixel_format == ILI9XXX_PIXEL_FORMAT_RGB565) {
 		pixel_format = PIXEL_FORMAT_RGB_565;
+	} else if (config->pixel_format == ILI9XXX_PIXEL_FORMAT_BGR565) {
```
RGb565 macro as well.

JarmouniA reviewed (commented):
Please squash fix-up commits, and reword commit messages to describe the actual changes.
The display mirroring fix should be in a separate commit.

NickIOT72 commented:
> Please squash fix-up commits, and reword commit messages to describe the actual changes. The display mirroring fix should be in a separate commit.

Ready. I left the first commit separated and squash the last 3 commits

JarmouniA reviewed (commented):
`include/zephyr/dt-bindings/display/ili9xxx.h` can be deleted.
Also, could you add an entry in the migration guide (`doc/releases/migration-guide-4.4.rst`) about this change (display mirroring, and the usage of standard pixel color macros in DT), so that users update their apps accordingly in the next release.

JarmouniA commented on doc/releases/migration-guide-4.4.rst:
```diff
@@ -32,6 +32,22 @@ Boards
 Device Drivers and Devicetree
 *****************************
 
+Display
```
In a separate commit please. Also, too much details and unrelated stuff here, we only want to inform the user about breaking changes that necessitate an action at their end and what that action is. Changes inside the driver are not relevant.

NickIOT72 replied:
Ready. I submit 3 commits: 1 for the original changes. other for the fix-up changes (overlay, yaml files and others) and the last one for the documentation

JarmouniA reviewed (commented):
Please delete `dt-bindings/display/ili9xxx.h`

JarmouniA reviewed (commented):
Last commit message title is incorrect.

NickIOT72 commented:
@JarmouniA All checks have passed

JarmouniA commented on boards/m5stack/m5stack_fire/m5stack_fire_procpu.dts:
```diff
@@ -95,7 +95,7 @@
 			mipi-max-frequency = <30000000>;
 			reg = <0>;
-			pixel-format = <ILI9XXX_PIXEL_FORMAT_RGB565>;
+			pixel-format = <0x20>;
```
> All of these files had the panel.h header but the macro wasn't recognized.

That's strange. Push a commit on top with the macros and I will have a look at the erros in CI.

JarmouniA replied:
The errors in Twister Build 2 are unrelated to the PR https://github.com/zephyrproject-rtos/zephyr/actions/runs/19791231302/job/56721806874?pr=99267
```
arch.arm.interrupt.no_optimizations on bl5340_dvk/nrf5340/cpuapp error (Build failure - collect2: error: ld returned 1 exit status)
INFO    - 2) arch.arm.swap.common.fpu_sharing.no_optimizations on bl5340_dvk/nrf5340/cpuapp error (Build failure - collect2: error: ld returned 1 exit status)
INFO    - 3) arch.arm.swap.common.no_optimizations on bl5340_dvk/nrf5340/cpuapp error (Build failure - collect2: error: ld returned 1 exit status)
INFO    - 4) arch.arm.interrupt.no_optimizations on bl5340_dvk/nrf5340/cpunet error (Build failure - collect2: error: ld returned 1 exit status)
```
The same for Build 3 https://github.com/zephyrproject-rtos/zephyr/actions/runs/19791231302/job/56721806879?pr=99267
```
arch.arm.swap.common.no_optimizations on bl5340_dvk/nrf5340/cpunet error (Build failure - collect2: error: ld returned 1 exit status)
```
The way forward is to check if there are already open issues regarding these failures, and comment on them to notify the people involved that the issues are CI Blocker so they get prioritized.
Otherwise create new bug issues for each test failure and I will mark them as CI blockers.

NickIOT72 committed:
doc: releases: add ili9xxx fixes to 4.4 releases notes & migration guide

Added documentation for doc/releases/migration-guide-4.4.rst
in reference with the changes on ili9xxx and ili9341
controllers

Signed-off-by: Nicolas Moreno <niko722795@gmail.com>

JarmouniA reviewed (dismissed):
Thanks!

JarmouniA commented:
@NickIOT72 Needs a rebase.
@VynDragon Could you take a look? Thanks!

VynDragon commented:
Ran on old cheap 2.4" ili9341 ihad laying around with parallel bitbang with esp32s3_devkitc clone, correct rendering as so:
```
/ {
	chosen {
		zephyr,display = &ili9341;
	};
	dbi_parallel {
		compatible = "zephyr,mipi-dbi-bitbang";
		status = "okay";
		clock-frequency= <100000>;
		#address-cells = <1>;
		#size-cells = <0>;

		reset-gpios = <&gpio0 4 GPIO_ACTIVE_LOW>;
		cs-gpios = <&gpio0 5 GPIO_ACTIVE_LOW>;
		dc-gpios = <&gpio0 6 GPIO_ACTIVE_HIGH>;
		wr-gpios = <&gpio0 7 GPIO_ACTIVE_HIGH>;
		rd-gpios = <&gpio0 15 GPIO_ACTIVE_HIGH>;
		data-gpios = <&gpio0 16 GPIO_ACTIVE_HIGH>,
			<&gpio0 17 GPIO_ACTIVE_HIGH>,
			<&gpio0 18 GPIO_ACTIVE_HIGH>,
			<&gpio0 8 GPIO_ACTIVE_HIGH>,
			<&gpio0 3 GPIO_ACTIVE_HIGH>,
			<&gpio1 14 GPIO_ACTIVE_HIGH>,
			<&gpio0 9 GPIO_ACTIVE_HIGH>,
			<&gpio0 10 GPIO_ACTIVE_HIGH>;
		ili9341: ili9341@0 {
			compatible = "ilitek,ili9341";
			mipi-max-frequency = <25000000>;
			reg = <0>;
			width = <240>;
			height = <320>;
			pixel-format = <PANEL_PIXEL_FORMAT_BGR_565>;
			disctrl = [0a C2 27 04];
			rotation = <0>;
			mipi-mode = "MIPI_DBI_MODE_8080_BUS_8_BIT";
		};
	};
};
```
![IMG_20260104_091103](https://github.com/user-attachments/assets/6585d630-ddaa-4ac9-ad10-5c71e6d6dd1f)
It needed disctrl changed because inverted gate scan order, switching between BGR and RGB produces expected and consistent results.
Setting orientation works as expected.

JarmouniA commented on boards/st/st25dv_mb1283_disco/st25dv_mb1283_disco.dts:
```diff
@@ -87,7 +87,7 @@
 			mipi-max-frequency = <DT_FREQ_M(20)>;
 			reg = <0>;
-			pixel-format = <ILI9XXX_PIXEL_FORMAT_RGB888>;
+			pixel-format = <PANEL_PIXEL_FORMAT_RGB_565>;
```
Ah missed this, @NickIOT72 can you keep the same format?

rexut reviewed (approved):
With this changeset, all LCD shields that are (still) maintained in Bridle and based on an Ilitek controller could be put back into operation with the basic display sample. Details about, see Bridle issue tiacsys/bridle#389 or PR tiacsys/bridle#390. This was also achieved by backporting to Zephyr v4.3:

- [GeeekPi Pico Breadboard Kit](https://bridle.tiac-systems.net/doc/4.3/bridle/boards/shields/rpi_pico_bb/doc/index.html#lcd-orientation-and-bit-order-test), `geeekpi_pico_bb` (**ILI9341**)
- [Waveshare 2.4 LCD](https://bridle.tiac-systems.net/doc/4.3/bridle/boards/shields/waveshare_lcd/doc/index.html#programming), `waveshare_2_4_lcd` (**ILI9341**)
- [Waveshare ResTouch LCD 3.5](https://bridle.tiac-systems.net/doc/4.3/bridle/boards/shields/rpi_pico_lcd/doc/index.html#lcd-orientation-and-bit-order-test), `waveshare_pico_restouch_lcd_3_5` (**ILI9488**)

Thank you very much for that contribution. We very much hope that the changes will soon be merged into `main` and possibly also reach the `v4.3-branch` via backporting.

JarmouniA commented:
> and possibly also reach the v4.3-branch via backporting.

We can backport the 1st commit (we only backport bug fixes).

kartben commented:
> Unfortunately I don't know why a backport was not created for this PR even though I added the backport label, but I will look into it again.

Probably a bug
https://github.com/zephyrproject-rtos/zephyr/actions/runs/24002906122/job/70001762865#step:2:148

JarmouniA commented:
> > Unfortunately I don't know why a backport was not created for this PR even though I added the backport label, but I will look into it again.
> 
> Probably a bug https://github.com/zephyrproject-rtos/zephyr/actions/runs/24002906122/job/70001762865#step:2:148

So if I understand it correctly, we are loosing a bunch of backport failure issues because the author of the to-be-backported PR is not a project contributor?
